File: node-server/src/metrics.ts
/**
 * Prometheus metrics (prom-client).
 *
 * Counterpart of `src/metrics.py`. Metric names, label sets and histogram
 * buckets are kept identical so existing dashboards/alerts work unchanged:
 *   mcp_tool_requests_total{tool,status}        Counter
 *   mcp_tool_duration_seconds{tool}             Histogram
 *   mcp_cache_operations_total{prefix,result}   Counter
 *   mcp_db_pool_size                            Gauge
 *   mcp_db_pool_checked_out                     Gauge
 *
 * Exposed on METRICS_PORT (default 9090) at /metrics, like the Python side.
 */

import http from "node:http";
import { Counter, Gauge, Histogram, register } from "prom-client";
import { config } from "./config.js";
import { logError, logInfo } from "./logger.js";

export const toolRequests = new Counter({
  name: "mcp_tool_requests_total",
  help: "Total MCP tool invocations",
  labelNames: ["tool", "status"] as const,
});

export const toolDuration = new Histogram({
  name: "mcp_tool_duration_seconds",
  help: "MCP tool execution latency",
  labelNames: ["tool"] as const,
  buckets: [0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1.0, 2.5, 5.0, 10.0],
});

export const cacheOps = new Counter({
  name: "mcp_cache_operations_total",
  help: "Redis cache hit/miss counts",
  labelNames: ["prefix", "result"] as const,
});

export const dbPoolSize = new Gauge({
  name: "mcp_db_pool_size",
  help: "Total connections in the pg pool",
});

export const dbPoolCheckedOut = new Gauge({
  name: "mcp_db_pool_checked_out",
  help: "Connections currently checked out from the pg pool",
});

/** Increment request counter and record latency for one tool invocation. */
export function recordToolCall(tool: string, status: string, durationS: number): void {
  toolRequests.labels(tool, status).inc();
  toolDuration.labels(tool).observe(durationS);
}

/** Increment the cache operations counter. result: "hit" | "miss" | "error". */
export function recordCacheOp(prefix: string, result: string): void {
  cacheOps.labels(prefix, result).inc();
}

/** Refresh DB pool gauges from a live pg Pool (totalCount / idleCount). */
export function updateDbPoolStats(pool: { totalCount: number; idleCount: number }): void {
  try {
    dbPoolSize.set(pool.totalCount);
    dbPoolCheckedOut.set(pool.totalCount - pool.idleCount);
  } catch {
    // ignore
  }
}

let _metricsServer: http.Server | null = null;

/** Start the Prometheus HTTP server. Idempotent (one bind per process). */
export function startMetricsServer(port?: number): number {
  const p = port ?? config().metricsPort;
  if (_metricsServer !== null) return p;
  const server = http.createServer(async (req, res) => {
    if (req.url === "/metrics") {
      try {
        const body = await register.metrics();
        res.writeHead(200, { "Content-Type": register.contentType });
        res.end(body);
      } catch (err) {
        res.writeHead(500);
        res.end(String((err as Error).message));
      }
      return;
    }
    res.writeHead(404);
    res.end();
  });
  server.on("error", (err: Error) => {
    logError(`Could not start metrics server on :${p}: ${err.message}`);
  });
  server.listen(p, () => {
    logInfo(`Prometheus metrics server started on :${p}/metrics`);
  });
  _metricsServer = server;
  return p;
}

/** Poll a pg-pool-shaped object and refresh gauges every `intervalMs`. */
export function startDbStatsCollector(
  getPool: () => { totalCount: number; idleCount: number },
  intervalMs = 15_000,
): NodeJS.Timeout {
  const timer = setInterval(() => {
    try {
      updateDbPoolStats(getPool());
    } catch {
      // ignore
    }
  }, intervalMs);
  timer.unref();
  return timer;
}

File: src/metrics.py
"""
Prometheus metrics.
Exposes a /metrics HTTP endpoint on METRICS_PORT (default 9090).

Usage:
    from metrics import record_tool_call, record_cache_op, start_metrics_server

Metrics exposed:
    mcp_tool_requests_total{tool, status}          Counter
    mcp_tool_duration_seconds{tool}                Histogram
    mcp_dependency_requests_total{dependency, operation, status}
    mcp_dependency_duration_seconds{dependency, operation}
    mcp_cache_operations_total{prefix, result}     Counter
    mcp_db_pool_size                               Gauge
    mcp_db_pool_checked_out                        Gauge
"""

import asyncio
import os
import time
from typing import Any, Callable

from prometheus_client import REGISTRY, Counter, Gauge, Histogram
from prometheus_client import start_http_server as _prom_start_http_server

from utils import log_error, log_info

# ── metric definitions ───────────────────────────────────────────────────────

tool_requests = Counter(
    "mcp_tool_requests_total",
    "Total MCP tool invocations",
    ["tool", "status"],
)

tool_duration = Histogram(
    "mcp_tool_duration_seconds",
    "MCP tool execution latency",
    ["tool"],
    buckets=[0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1.0, 2.5, 5.0, 10.0],
)

cache_ops = Counter(
    "mcp_cache_operations_total",
    "Redis cache hit/miss counts",
    ["prefix", "result"],  # result: hit | miss | error
)

dependency_requests = Counter(
    "mcp_dependency_requests_total",
    "External/internal dependency calls by dependency, operation, and status",
    ["dependency", "operation", "status"],
)

dependency_duration = Histogram(
    "mcp_dependency_duration_seconds",
    "Dependency call latency by dependency and operation",
    ["dependency", "operation"],
    buckets=[
        0.001,
        0.0025,
        0.005,
        0.01,
        0.025,
        0.05,
        0.1,
        0.25,
        0.5,
        1.0,
        2.5,
        5.0,
        10.0,
        30.0,
    ],
)

db_pool_size = Gauge(
    "mcp_db_pool_size",
    "Total connections in the asyncpg pool",
)

db_pool_checked_out = Gauge(
    "mcp_db_pool_checked_out",
    "Connections currently checked out from the asyncpg pool",
)


# ── helpers called from other modules ────────────────────────────────────────


def record_tool_call(tool: str, status: str, duration_s: float) -> None:
    """Increment request counter and record latency for one tool invocation.

    Args:
        tool: MCP tool name (e.g. ``"search_medical_codes"``).
        status: Outcome — ``"success"`` or ``"error"``.
        duration_s: Wall-clock execution time in seconds.
    """
    tool_requests.labels(tool=tool, status=status).inc()
    tool_duration.labels(tool=tool).observe(duration_s)


def record_cache_op(prefix: str, result: str) -> None:
    """Increment the cache operations counter for one cache event.

    Args:
        prefix: Cache namespace / prefix string.
        result: Outcome — ``"hit"``, ``"miss"``, or ``"error"``.
    """
    cache_ops.labels(prefix=prefix, result=result).inc()


def record_dependency_call(
    dependency: str,
    operation: str,
    status: str,
    duration_s: float,
) -> None:
    """Record latency and outcome for a DB/cache/HTTP/embedding dependency call.

    Keep labels intentionally low-cardinality: ``operation`` should be a stable
    method/stage name, never raw SQL, URLs, user input, or IDs.
    """
    dependency_requests.labels(
        dependency=dependency,
        operation=operation,
        status=status,
    ).inc()
    dependency_duration.labels(
        dependency=dependency,
        operation=operation,
    ).observe(duration_s)


def update_db_pool_stats(pool: Any) -> None:
    """Refresh DB pool Prometheus gauges from a live asyncpg Pool object.

    Args:
        pool: An ``asyncpg.Pool`` instance (typed as ``Any`` to avoid a
            hard dependency on asyncpg in this module).
    """
    try:
        db_pool_size.set(pool.get_size())
        db_pool_checked_out.set(pool.get_size() - pool.get_idle_size())
    except Exception:
        pass


# ── periodic DB stats collection ────────────────────────────────────────────


async def _collect_db_stats_loop(get_pool_fn: Callable, interval: int = 15) -> None:
    while True:
        try:
            pool = get_pool_fn()
            update_db_pool_stats(pool)
        except Exception:
            pass
        await asyncio.sleep(interval)


# ── server startup ────────────────────────────────────────────────────────────

_metrics_server_started = False


def start_metrics_server(port: int | None = None) -> int:
    """Start the Prometheus HTTP server. Idempotent across FastMCP session lifespans.

    Args:
        port: Port to bind. Defaults to the ``METRICS_PORT`` env var, then ``9090``.

    Returns:
        The port actually used.
    """
    global _metrics_server_started
    port = port or int(os.getenv("METRICS_PORT", "9090"))
    if _metrics_server_started:
        return port
    try:
        _prom_start_http_server(port)
        _metrics_server_started = True
        log_info(f"Prometheus metrics server started on :{port}/metrics")
    except OSError as e:
        log_error(f"Could not start metrics server on :{port}: {e}")
    return port


async def start_db_stats_collector(
    get_pool_fn: Callable, interval: int = 15
) -> asyncio.Task:
    """Launch a background task that refreshes DB pool Prometheus gauges.

    Args:
        get_pool_fn: Zero-argument callable that returns the current asyncpg Pool.
        interval: Polling interval in seconds.

    Returns:
        The running ``asyncio.Task``.
    """
    return asyncio.create_task(_collect_db_stats_loop(get_pool_fn, interval))
